Add EIP: Block-level Warming with fair cost savings

forshtat opened the pull request:
Reopening the old PR that I did not realise was not previously merged (https://github.com/ethereum/EIPs/pull/7968)

This proposal is significantly different from EIP-7863 (https://github.com/ethereum/EIPs/pull/9246), arguably presents a more complete and fair solution that is worth the extra complexity, and precedes it for more than a year.

bomanaps commented on EIPS/eip-7557.md:
```diff
@@ -0,0 +1,375 @@
+accounts and storage slots that the transaction plans to access,
+however, the cost is still paid repeatedly by each transaction rather than once at the block level.
+
+With the [EIP-6800: Ethereum state using a unified verkle tree](./eip-6800) on the roadmap for inclusion,
+the cost of reading from the Ethereum state and especially the contract code is expected to increase.\
```
```suggestion
With the [EIP-6800: Ethereum state using a unified verkle tree](./eip-6800) on the roadmap for inclusion, the cost of reading from the Ethereum state and especially the contract code is expected to increase.
```

bomanaps commented on EIPS/eip-7557.md:
```diff
@@ -0,0 +1,375 @@
+do not accurately reflect the computations that block builders and validators perform for the blockchain state access
+during transaction execution.
+
+[EIP-2930: Optional access lists](./eip-2930) made it possible for transactions to pre-specify and pre-pay for the
+accounts and storage slots that the transaction plans to access,
+however, the cost is still paid repeatedly by each transaction rather than once at the block level.
```
```suggestion
[EIP-2930: Optional access lists](./eip-2930) made it possible for transactions to pre-specify and pre-pay for the accounts and storage slots that the transaction plans to access, however, the cost is still paid repeatedly by each transaction rather than once at the block level.
```

bomanaps commented on EIPS/eip-7557.md:
```diff
@@ -0,0 +1,375 @@
+between "cold" and "warm" access to accounts and storage slots.\
+However, the cost of every cold access is borne by each transaction separately, even though the validator only
+needs to fetch the state object once for the entire block.\
+When multiple transactions access the same state object in the same block the fees charged for these transactions
+do not accurately reflect the computations that block builders and validators perform for the blockchain state access
+during transaction execution.
```
```suggestion
When multiple transactions access the same state object in the same block the fees charged for these transactions do not accurately reflect the computations that block builders and validators perform for the blockchain state access during transaction execution.
```

bomanaps commented on EIPS/eip-7557.md:
```diff
@@ -0,0 +1,375 @@
+
+[EIP-2929: Gas cost increases for state access opcodes](./eip-2929) introduced a new gas cost model that differentiates
+between "cold" and "warm" access to accounts and storage slots.\
+However, the cost of every cold access is borne by each transaction separately, even though the validator only
+needs to fetch the state object once for the entire block.\
```
```suggestion
However, the cost of every cold access is borne by each transaction separately, even though the validator only needs to fetch the state object once for the entire block.
```

bomanaps commented on EIPS/eip-7557.md:
```diff
@@ -0,0 +1,375 @@
+
+## Motivation
+
+[EIP-2929: Gas cost increases for state access opcodes](./eip-2929) introduced a new gas cost model that differentiates
+between "cold" and "warm" access to accounts and storage slots.\
```
```suggestion
[EIP-2929: Gas cost increases for state access opcodes](./eip-2929) introduced a new gas cost model that differentiates between "cold" and "warm" access to accounts and storage slots.
```

bomanaps commented on EIPS/eip-7557.md:
```diff
@@ -0,0 +1,375 @@
+
+## Abstract
+
+A mechanism for a fair distribution of the gas costs associated with access to addresses and storage slots
+among multiple transactions with shared items in their `accessList`.
```
```suggestion
A mechanism for a fair distribution of the gas costs associated with access to addresses and storage slots among multiple transactions with shared items in their `accessList`.
```

bomanaps commented on EIPS/eip-7557.md:
```diff
@@ -0,0 +1,375 @@
+
+The contents of the `accessList` parameter are part of the Ethereum history and the potential cost of keeping this
+data in the blockchain must be accounted for when implementing this change.
+There is currently no additional charge applied to the `accessList` parameter, due to the cost of including
+an address or a storage slot in the `accessList` being a constant value that is significantly higher than the
+potential cost of storing the `accessList` at the cost of  a
+dynamically sized `calldata` field.
```
```suggestion
There is currently no additional charge applied to the `accessList` parameter, due to the cost of including an address or a storage slot in the `accessList` being a constant value that is significantly higher than the potential cost of storing the `accessList` at the cost of  a dynamically sized `calldata` field.
```

bomanaps commented on EIPS/eip-7557.md:
```diff
@@ -0,0 +1,375 @@
+
+A common game-theoretical answer to the problem of calculating a fair redistribution of the payoff of the
+results of the participants' cooperation is the use of Shapley values.\
+However, we argue that the proposed distribution of the `priorityFee` reimbursements is sufficiently fair while being
+a lot easier to compute or articulate.
```
```suggestion
However, we argue that the proposed distribution of the `priorityFee` reimbursements is sufficiently fair while being a lot easier to compute or articulate.
```

bomanaps commented on EIPS/eip-7557.md:
```diff
@@ -0,0 +1,375 @@
+
+## Security Considerations
+
+The upper limit of storage reads in one block is not affected by this change as the gas charge is done with the
+full cost of `COLD_ACCOUNT_ACCESS_COST` or `COLD_SLOAD_COST`.
```
```suggestion
The upper limit of storage reads in one block is not affected by this change as the gas charge is done with the full cost of `COLD_ACCOUNT_ACCESS_COST` or `COLD_SLOAD_COST`.
```

bomanaps commented on EIPS/eip-7557.md:
```diff
@@ -0,0 +1,375 @@
+The upper limit of storage reads in one block is not affected by this change as the gas charge is done with the
+full cost of `COLD_ACCOUNT_ACCESS_COST` or `COLD_SLOAD_COST`.
+
+The maximum amount of memory and computation required to calculate the reimbursements according to the
+specified algorithm is insignificant.
```
```suggestion
The maximum amount of memory and computation required to calculate the reimbursements according to the specified algorithm is insignificant.
```

bomanaps reviewed (commented):
Please kindly take a look at the lining format. Usually, you start a new sentence on a new line.

g11tech commented on EIPS/eip-7557.md:
```diff
@@ -0,0 +1,388 @@
+
+### Calculating a reimbursement of the charged priority fee
+
+Each transaction pays an individual `priorityFeePerGas` value and redistributing this part of the cold access cost
```
each tx pays priority + base fee, you only applying warming on priority fee?

forshtat replied:
Yes, as the base fee is same for all transactions in a block it is easy to just split the cost. Priority fee is different and we need some kind of a formula to redistribute its costs.

forshtat replied:
Basically, we reimburse the base fee by redistributing all the base fee collected to all transactions accessing the storage, this is described in `### Calculating a reimbursement of the burned base fee`. Maybe you could suggest a better way to define this behaviour?

forshtat commented:
Hello @bomanaps and thank you for the review. I have removed the use of `\` as a line break, and fixed a few long lines containing multiple sentences.
However, I am not sure I can agree with your suggestion of using the newline **only** to separate sentences. As sentences can become very long in these technical documents and especially when using `git` to manage versions of these documents, long lines create an inconvenience while having absolutely no observable effect on the rendered markdown file.
Please advise if this is a strong requirement for the EIPs to be merged going forward or if it is just a suggestion.
Thanks.

SamWilsn commented on EIPS/eip-7557.md:
```diff
@@ -0,0 +1,403 @@
+---
+eip: 7557
+title: Block-level Warming with fair cost savings
+description: Block-level warming of addresses and slots with access lists
```
You have a bit more room (144 characters) in your description. You should expand and give a bit more detail.

SamWilsn commented on EIPS/eip-7557.md:
```diff
@@ -0,0 +1,403 @@
+
+## Abstract
+
+A mechanism for a fair distribution of the gas costs associated with access to addresses and storage slots
+among multiple transactions with shared items in their `accessList`.
```
The abstract should contain a brief (but still technical) overview of the proposal. Can you sketch out the mechanism here without getting too in the weeds?

SamWilsn reviewed (approved):
This is good to go for a draft. Do address these comments before moving into Review.
